src/validation.cpp

```diff
@@ -4550,22 +4771,46 @@ void CChainState::CheckBlockIndex(const Consensus::Params& consensusParams)
-                // is valid and we have all data for its parents, it must be in
-                // setBlockIndexCandidates.  chainActive.Tip() must also be there
-                // even if some data has been pruned.
-                if (pindexFirstMissing == nullptr || pindex == chainActive.Tip()) {
-                    assert(setBlockIndexCandidates.count(pindex));
-                }
-                // If some parent is missing, then it could be that this block was in
-                // setBlockIndexCandidates but had to be removed because of the missing data.
-                // In this case it must be in mapBlocksUnlinked -- see test below.
+        if (pindex->nStatus & BLOCK_FAILED_CHILD) {
+            // Blocks which failed with "CHILD" must have an invalid parent
+            assert(pindexFirstInvalid);
+            assert(pindexFirstInvalid != pindex);
+            assert(!(pindex->nStatus & BLOCK_FAILED_VALID));
+        }
+        if (pindex->nStatus & BLOCK_FAILED_VALID) {
+            assert(!(pindex->nStatus & BLOCK_FAILED_CHILD));
+        }
+        if (!CBlockIndexWorkComparator()(pindex, chainActive.Tip()) && pindexFirstNeverProcessed == nullptr && pindexFirstInvalid == nullptr) {
+            // If this block sorts at least as good as the current tip and
+            // is valid and we have all data for its parents, it must be in
+            // setBlockIndexCandidates.  chainActive.Tip() must also be there
+            // even if some data has been pruned.
+            if (pindexFirstMissing == nullptr || pindex == chainActive.Tip()) {
+                assert(setBlockIndexCandidates.count(pindex));
             }
-        } else { // If this block sorts worse than the current tip or some ancestor's block has never been seen, it cannot be in setBlockIndexCandidates.
+            // If some parent is missing, then it could be that this block was in
+            // setBlockIndexCandidates but had to be removed because of the missing data.
+            // In this case it must be in mapBlocksUnlinked -- see test below.
+        } else { // If this block sorts worse than the current tip or some ancestor's block has never been seen or is invalid, it cannot be in setBlockIndexCandidates.
             assert(setBlockIndexCandidates.count(pindex) == 0);
         }
+        bool must_be_header_candidate_if_leaf = false;
+        if (pindex->nChainWork < chainActive.Tip()->nChainWork) {
+            // Irrespective of pruned-ness, if this block sorts worse than the current tip, it cannot be in setBlockIndexHeaderCandidates
+            assert(setBlockIndexHeaderCandidates.count(pindex) == 0);
+        } else if (pindexFirstInvalid == nullptr) {
+            // If pindex is a leaf node and sorts at the same or greater height
+            // than chainActive.Tip(), it must be in setBlockIndexHeaderCandidates.
+            must_be_header_candidate_if_leaf = true;
+        }
+        if (setBlockIndexHeaderCandidates.count(pindex)) {
```

Comment:
`setBlockIndexHeaderCandidates.count(pindex) > 0` to improve readability and avoid implicit conversion :-)